[zienxu] iP

zienxu opened the pull request:
# Sammy Chatbot
>Stay organised, stay ahead. - me

- Text-based interface
- Date parsing support
- Organizes tasks into a list

## How to use
1. Download the jar file [here](https://github.com/zienxu/ip/releases/tag/A-Jar).
2. Download java 17 [here](https://www.oracle.com/sg/java/technologies/downloads/#java17).
3. Run java -jar ip.jar on your command line!
4. Use the chatbot as you wish.

## Features

- [x] Supports TODO, Deadline, and Event tasks

- [ ] Reminders (upcoming)

- [ ] UI Improvements (upcoming)

## Java code example used in the programme
```
package sammy.command;

import sammy.Storage;
import sammy.task.TaskList;
import sammy.Ui;

public class ListCommand extends Command {
    @Override
    public void execute(TaskList tasks, Ui ui, Storage storage) {
        ui.showTaskList(tasks);
    }
}
```

Commit:
Add Gradle support

Eclipse-Dominator committed:
Bump gradle and lib version

aureliony committed:
build.gradle: Prevent generating a second JAR file

In build.gradle, the dependencies on distZip and/or distTar causes
the shadowJar task to generate a second JAR file for which the
mainClass.set("seedu.duke.Duke") does not take effect.
Hence, this additional JAR file cannot be run.
For this product, there is no need to generate a second JAR file
to begin with.

Let's remove this dependency from the build.gradle to prevent the
shadowJar task from generating the extra JAR file.

zienxu committed:
add Sammy.java

zienxu committed:
Added level 1 requirements to Sammy.java

zienxu committed:
added level 2 requirements to Sammy.java

zienxu committed:
added level 3 requirements to Sammy.java

zienxu committed:
added level 4 requirements to Sammy.java

zienxu committed:
added level 5 requirements to Sammy.java

zienxu committed:
Added level 6 requirements to Sammy.java and Automated Text UI Testing

zienxu committed:
complete level 7 requirements

The current implementation does not save tasks to disk, so all data is lost when the chatbot restarts. This is inconvenient for users who need to maintain a persistent task list.

This commit adds functionality to automatically save the task list to a file (`./data/sammy.txt`) whenever it changes. It also loads the task list from the file when the chatbot starts. This ensures that tasks are preserved across sessions.

Let's handle cases where the file or folder doesn't exist by creating them when necessary. File paths are specified in an OS-independent way to ensure cross-platform compatibility.

zienxu committed:
Level-8: Add support for parsing and storing dates and times

Teach the chatbot to understand and parse dates and times for tasks.
Store deadlines as `LocalDateTime` in task objects and accept various
date formats such as `yyyy-MM-dd HHmm` or `d/M/yyyy HHmm`. Display
dates in the format `MMM dd yyyy, h:mm a` for better readability.

This commit enables the chatbot to interpret and handle date-specific
commands, enhancing its capability to manage tasks with deadlines
effectively.

zienxu committed:
Refactor code to improve OOP design by extracting related functionality into classes

The existing codebase had multiple responsibilities bundled together, making it harder to maintain and extend. Specifically, user interaction, file handling, and task management were handled within a single class.

To improve the code quality and adhere to OOP principles, Ui, Storage, Parser, and TaskList Classes were added.

Additionally, Command classes were introduced to represent different actions (e.g., AddCommand, DeleteCommand, ExitCommand), allowing the main application loop to handle commands in a more modular and maintainable way.

zienxu committed:
Divide classes into suitable packages

To improve code organisation and maintainability, the project structure
has been refactored to place classes into appropriate Java packages.

Moved classes into their respective packages:
  - 'sammy.command' for command-related classes
  - 'sammy.task' for task-related classes
  - 'sammy' for core classes such as 'Sammy', 'Storage', 'TaskList', etc.

This change helps in better structuring the codebase as the project
grows.

zienxu committed:
Merge remote-tracking branch 'origin/add-gradle-support'

Merge Gradle to the master branch

It is a requirement by the course.
Gradle support is provided as a separate branch named add-gradle-support in the Duke repo.

zienxu committed:
Merge branch 'branch-Level-8'

Merging branch level-8 to the master branch.

zienxu committed:
Merge branch 'branch-A-MoreOOP'
Merge A-MoreOOP with master branch

zienxu committed:
Merge branch 'branch-A-Packages'
Merge branch A-Packages with master branch

JumpyJay commented on src/main/java/sammy/task/Event.java:
```diff
@@ -0,0 +1,27 @@
+
+    public Event(String description, String startTime, String endTime) {
+        super(description);
+        DateTimeFormatter inputFormat = DateTimeFormatter.ofPattern("yyyy-MM-dd HHmm");
```
Maybe could change to 'inputFormatter' to better align with the Cody Quality: Naming conventions

JumpyJay commented on src/main/java/sammy/Parser.java:
```diff
@@ -0,0 +1,84 @@
+
+public class Parser {
+    public static Command parse(String fullCommand) throws SammyException {
+        String[] parts = fullCommand.split(" ", 2);
```
Could maybe rename to commandParts for better clarity

JumpyJay commented on src/main/java/sammy/Parser.java:
```diff
@@ -0,0 +1,84 @@
+    }
+
+    public static Task parseTask(String line) {
+        String[] parts = line.split(" \\| ");
```
Similarly, could consider changing the wording to 'taskParts'

JumpyJay commented on src/main/java/sammy/Sammy.java:
```diff
@@ -0,0 +1,45 @@
+
+    public void run() {
+        ui.showWelcomeMessage();
+        boolean isExit = false;
```
Unlike the isExit boolean in the Command class which is easily understood, the isExit boolean here does not have the context of being related to a command. Would be helpful if the wording could be change to something along the lines of 'isExitCommandReceived()'

JumpyJay reviewed (commented):
LGTM, just some nits to fix :3

zienxu committed:
Implement JUnit requirements

haleemairfan commented on src/main/java/sammy/SammyException.java:
would you want to shift this to the sammy.exceptions package for better clarity. Also would be great to add javadoc comments to this file so readers can understand what a SammyException actually is!

haleemairfan reviewed (commented):
Hi! Your code looks good :) But for better readability you could add javadoc comments to your files and rename some variables!

zienxu committed:
Implement Level 10 requirements

zienxu committed:
Add assertion to AddCommand

zienxu committed:
Add assertion to ExitCommand

zienxu committed:
Add assertion to UnmarkCommand

zienxu committed:
Test pull request for CodeQuality

zienxu committed:
Command Classes: added assertions for Command Classes

Command classes now include assertions to enforce preconditions and
ensure internal consistency.

Assertions have been added to check for the validity of inputs and
states, thereby providing a safeguard against invalid data and potential
bugs. This change helps in identifying issues early in the development
process.

By asserting preconditions, we can ensure that invalid states are
avoided before any complex logic is executed in the command classes.

The addition of assertions ensures that the system behaves predictably
when used as intended, while providing clear error messages when it
doesn't. This is crucial for avoiding subtle bugs that can occur due
to misuse of commands.

Using assertions also allows us to maintain clearer and more reliable
code. These assertions focus on validating critical command conditions
rather than relying solely on defensive programming patterns.

zienxu committed:
add assertions for Task Sub-classes

zienxu committed:
add assertions to Parser, Sammy, Storage and UI Classes

zienxu committed:
Merge pull request #2 from zienxu/branch-A-Assertions

Add assertions to codebase

zienxu committed:
Add code quality requirements to Parser, Sammy, Storage and UI Classes

zienxu committed:
Merge branch 'master' into branch-A-CodeQuality

zienxu committed:
Merge pull request #3 from zienxu/branch-A-CodeQuality

Refine code quality

zienxu committed:
added jar file requirements to build gradle

zienxu committed:
add tag class TagCommand

zienxu committed:
testing Tag class

zienxu committed:
fix | Tags: being printed repeatedly when program reruns problem

zienxu committed:
Merge pull request #4 from zienxu/branch-BCD-Extension

Add BCD Extension requirement of Tagging

zienxu committed:
Add TextUiTesting requirement

zienxu committed:
Merge pull request #5 from zienxu/branch-A-TextUiTesting

Add TextUiTesting requirement

zienxu committed:
Add UI.png

zienxu committed:
Add UI.png correctlyto docs folder

zienxu committed:
Add user guide but work in progress

zienxu committed:
update user guide

zienxu committed:
Edit User Guide, added Quick Start

zienxu committed:
Edit build.gradle

zienxu committed:
Merge pull request #6 from zienxu/branch-A-UserGuide

Add user guide

zienxu committed:
Update README.md

zienxu committed:
Update README.md

zienxu committed:
Update README.md
